### PR TITLE
[RPD-192] Create a `destroy` function in core

### PR DESCRIPTION
**Pull request by swells2020**
This PR removes the `destroy.py` module from `matcha-ml.cli` and moves the functionality into `core.py`. The logic in core can now be accessed from the cli by utilising the `matcha-ml.core` module.

The UI logic from `destroy.py` has been added to the `destroy` function in the `cli.py` module. The functional logic from `destroy.py` has been added to the `core.py` module. The errors have been updated.

A duplicative UI message was removed from `remote_state_manager.RemoteStateManager.deprovision_remote_state`.

The tests in `test_cli.test_destroy` have been updated.

Tests were added to`core.destroy`.

## Checklist

Please ensure you have done the following:

* [X] I have read the [CONTRIBUTING](https://github.com/fuzzylabs/matcha/blob/main/CONTRIBUTING.md) guide.
* [ ] I have updated the documentation if required.
* [X] I have added tests which cover my changes.

## Type of change

Tick all those that apply:

* [ ] Bug Fix (non-breaking change, fixing an issue)
* [ ] New feature (non-breaking change to add functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [X] Other (add details above)

### CODE REVIEW

File: src/matcha_ml/core/core.py [hunk 1]
```diff
@@ -76,6 +83,41 @@ def get(
     return result
 
 
+def destroy(resources: List[Tuple[str, str]]) -> None:
+    """Destroy resources.
```
Comment (JonoCX) [inline src/matcha_ml/core/core.py:87]: A slightly more descriptive docstring here would be good

File: src/matcha_ml/core/core.py [hunk 2]
```diff
@@ -76,6 +83,41 @@ def get(
...
+
+    Args:
+        resources (List[Tuple[str,str]): the list of resources to be actioned by the verb to be provided to the user as a status message
```
Comment (JonoCX) [inline src/matcha_ml/core/core.py:90]: Why does the API `destroy` function **require** a parameter to construct a UI message? This should all be done in the `cli` prior to calling this function.

**Review by JonoCX — CHANGES_REQUESTED** [review]
Nice work 🦾  See comments - the UI function calls need to be stripped out of `core.destroy`.

> I think the core.destroy functions needs tests adding.

Yep, we need tests for it.

File: src/matcha_ml/core/core.py [hunk 3]
```diff
@@ -76,6 +83,41 @@ def get(
...
+
+    Raises:
+        Matcha Error: where no state has been provisioned.
+    """
+    remote_state_manager = RemoteStateManager()
+    if not remote_state_manager.is_state_provisioned():
+        print_error(
```
[the remark this answers is not in the document]
Comment (swells2020) [reply src/matcha_ml/core/core.py]: Updated, thanks Jon.

File: src/matcha_ml/core/core.py (hunk 1 above)
Comment (swells2020) [reply src/matcha_ml/core/core.py:84]: Updated to match the cli `destroy --help` description.

File: src/matcha_ml/core/core.py [hunk 4]
```diff
@@ -76,6 +83,41 @@ def get(
...
+        )
+        raise MatchaError(
+            "Error - resources that have not been provisioned cannot be destroyed. Run 'matcha provision' to get started!"
+        )
+
+    with remote_state_manager.use_lock(), remote_state_manager.use_remote_state():
+        # create a runner for deprovisioning resource with Terraform service.
+        template_runner = AzureRunner()
+
+        if is_user_approved(verb="destroy", resources=resources):
```
[the remark this answers is not in the document]
Comment (swells2020) [reply src/matcha_ml/core/core.py]: Removed, thanks!

File: src/matcha_ml/core/core.py (hunk 2 above)
Comment (swells2020) [reply src/matcha_ml/core/core.py]: Removed, thanks!

File: src/matcha_ml/core/core.py [hunk 5]
```diff
@@ -76,6 +83,41 @@ def get(
...
+            # deprovision the resources
```
[the remark this answers is not in the document]
Comment (swells2020) [reply src/matcha_ml/core/core.py]: Removed, thanks!

File: src/matcha_ml/core/core.py [hunk 6]
```diff
@@ -83,6 +83,26 @@ def get(
     return result
 
 
+@track(event_name=AnalyticsEvent.DESTROY)
+def destroy() -> None:
+    """Destroys provisioned resources in the cloud.
```
Comment (JonoCX) [inline src/matcha_ml/core/core.py:88]: Nit: Thinking about the API documentation, this could be more detailed.

**Review by JonoCX — APPROVED** [review]
LGTM - nice work 🚀

**Review by Christopher-Norman — APPROVED** [review]
LGTM, one nit 👍

File: tests/test_services/test_analytics_service.py [hunk 7]
```diff
@@ -1,13 +1,14 @@
...
+from unittest.mock import MagicMock, PropertyMock, patch
 
 import pytest
 
 from matcha_ml.cli.cli import app
 from matcha_ml.services.global_parameters_service import GlobalParameters
 
-GLOBAL_PARAMETER_SERVICE_FUNCTION_STUB = (
+CORE_FUNCTION_STUB = "matcha_ml.core.core"
+GLOBAL_PARAMETER_SERVICE_CORE_FUNCTION_STUB = (
```
[the remark this answers is not in the document]
Comment (swells2020) [reply tests/test_services/test_analytics_service.py:11]: I think so, tests fail without it.

[the remark this answers is not in the document]
Comment (Christopher-Norman) [reply tests/test_services/test_analytics_service.py:11]: Sorry, I mean does the name need to include `CORE` in `GLOBAL_PARAMETER_SERVICE_CORE_FUNCTION_STUB`

[the remark this answers is not in the document]
Comment (swells2020) [reply tests/test_services/test_analytics_service.py:11]: My apologies. No, it doesn't. Good spot! Thanks, I'll update.

**Pull request merged**